File: CSharp/Step2/SftpActors.cs
using System;
using System.Collections.Generic;
using System.Linq;
using Akka.Actor;

using Shared;

namespace SftpActors
{
	public class SftpActor : ReceiveActor
	{
		private IClientFactory _clientFactory;
		private ISftpClient _connection;

		public SftpActor(IClientFactory clientFactory)
		{
			_clientFactory = clientFactory;

			Disconnected();
		}

		public class Connect { }
		public class Disconnect { }
		public class ListDirectory
		{
			public ListDirectory(string remotePath)
			{
				this.RemotePath = remotePath;
			}

			public string RemotePath { get; private set; }
		}

		private void Disconnected()
		{
			Receive<Connect>((cmd) =>
			{
				_connection = _clientFactory.CreateSftpClient();
				_connection.Connect();

				Become(Connected);
			});
		}

		private void Connected()
		{
			Receive<ListDirectory>((cmd) =>
			{
				IEnumerable<SftpFileInfo> result = null;
				try
				{
					result = _connection.ListDirectory(cmd.RemotePath, null);
				}
				catch (Exception)
				{
					result = new SftpFileInfo[] { };
				}
				this.Sender.Tell(result);
			});
			Receive<Disconnect>((cmd) =>
			{
				_connection.Disconnect();
				_connection.Dispose();

				Become(Disconnected);
			});
		}
	}
}

File: FSharp/Step2/SftpActors.fs
[<AutoOpen>]
module SftpActors
    open System
    open Akka.FSharp
    open SftpClient

    type SftpCommand =
        | Connect
        | Disconnect
        | ListDirectory of Url

    let sftpActor (clientFactory : IClientFactory) (mailbox: Actor<_>) =

        let rec disconnected () = 
            actor {
                let! message = mailbox.Receive ()
                match message with
                | Connect -> 
                    let connection = clientFactory.CreateSftpClient()
                    connection.Connect()
                    return! connected (connection)

                | _ ->
                    cprintfn ConsoleColor.Yellow "Sftp: invalid operation in disconnected state: %A" message

                return! disconnected ()
            } 
        and connected (connection) = 
            actor {
                let! message = mailbox.Receive ()
                match message with
                | Disconnect -> 
                    connection.Disconnect()
                    connection.Dispose()
                    return! disconnected ()

                | ListDirectory remotePath -> 
                    let result = 
                        try
                            connection.ListDirectory(remotePath.Value, noProgressCallback) 
                            |> List.ofSeq
                        with
                        | ex -> List.empty
                    mailbox.Sender() <! result

                | _ ->
                    cprintfn ConsoleColor.Yellow "Sftp: invalid operation in connected state: %A" message

                return! connected (connection)
            } 

        disconnected ()
